### PR TITLE
STAC-5310 push docker image with latest tag

### PR DESCRIPTION
**Pull request by johnnygladwin1991**
### What does this PR do?

Add the trace agent docker image

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -285,7 +289,7 @@ build_rpm:
   - inv -e agent.omnibus-build --base-dir /omnibus --skip-sign
 
     # Prepare outputs
-  - mkdir -p $CI_PROJECT_DIR/outcomes/pkg && mkdir -p $CI_PROJECT_DIR/outcomes/dockerfiles
+  - mkdir -p $CI_PROJECT_DIR/outcomes/pkg && mkdir -p $CI_PROJECT_DIR/outcomes/dockerfiles
```
Comment (TizianoPerrucci) [inline .gitlab-ci.yml]: remove trailing space

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -407,6 +411,21 @@ pre_release_image:
     - export VERSION=`cat version.txt`
     - ./omnibus/package-scripts/publish_image.sh $CI_COMMIT_REF_NAME $STS_DOCKER_TEST_REPO $CI_PROJECT_DIR/outcomes/pkg
 
+pre_release_trace_agent:
```
Comment (TizianoPerrucci) [inline .gitlab-ci.yml]: call it `pre_release_trace_agent_image`

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -407,6 +411,21 @@ pre_release_image:
...
+  stage: pre_release
+  image: docker:18.06.0
+  dependencies: 
+    - deps_deb
```
Comment (TizianoPerrucci) [inline .gitlab-ci.yml]: you don't need `deps_deb` if you override the before scripts with `before_script: []`

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -407,6 +411,21 @@ pre_release_image:
...
+    - build_deb
+  variables:
+    DOCKER_DRIVER: overlay
+  services:
+    - docker:dind
+  script:
+    - cd $CI_PROJECT_DIR
+    - export VERSION=`cat version.txt`
+    - ./omnibus/package-scripts/publish_trace_agent.sh $CI_COMMIT_REF_NAME $STS_DOCKER_TEST_REPO $CI_PROJECT_DIR/outcomes/binary/trace-agent
```
Comment (TizianoPerrucci) [inline .gitlab-ci.yml]: don't use `$STS_DOCKER_TEST_REPO` instead define a variable at the top like the other `STS_DOCKER_TEST_REPO_TRACE: stackstate-trace-agent-test` and use it here

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -496,6 +515,7 @@ release_image:
   stage: release
   image: docker:18.06.0
   dependencies:
+    - deps_deb
```
Comment (TizianoPerrucci) [inline .gitlab-ci.yml]: this can be remove

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -512,6 +532,26 @@ release_image:
...
+  variables:
+    DOCKER_DRIVER: overlay
+  services:
+    - docker:dind
+  
+  script:
+    - cd $CI_PROJECT_DIR
+    - export VERSION=`cat version.txt`
+    - export PUSH_LATEST=true
+    - ./omnibus/package-scripts/publish_trace_agent.sh $VERSION $STS_DOCKER_RELEASE_REPO $CI_PROJECT_DIR/outcomes/binary/trace-agent $PUSH_LATEST
```
Comment (TizianoPerrucci) [inline .gitlab-ci.yml]: don't use `$STS_DOCKER_RELEASE_REPO` instead define a variable at the top like the other `STS_DOCKER_RELEASE_REPO_TRACE: stackstate-trace-agent` and use it

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -512,6 +532,26 @@ release_image:
...
+  stage: release
+  image: docker:18.06.0
+  dependencies:
+    - deps_deb
+    - build_deb
+  variables:
+    DOCKER_DRIVER: overlay
+  services:
+    - docker:dind
+
```
Comment (TizianoPerrucci) [inline .gitlab-ci.yml]: remove new line

File: .gitlab-ci.yml [hunk 8]
```diff
@@ -512,6 +532,26 @@ release_image:
     - tags
     - triggers
 
+release_trace-agent:
+  stage: release
+  image: docker:18.06.0
+  dependencies:
+    - deps_deb
```
Comment (TizianoPerrucci) [inline .gitlab-ci.yml]: you don't need deps_deb if you override the before scripts with `before_script: []`

File: Dockerfiles/trace-agent/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+# - Remove the /var/run -> /run symlink and create a legit /var/run folder
+# as some docker versions re-create /run from zero at container start
+
+RUN  adduser --system --no-create-home --disabled-password --ingroup root stackstate-agent \
+ && chown -R stackstate-agent:root /etc/stackstate-agent/ /var/log/stackstate-agent/ \
+ && chmod g+r,g+w,g+X -R /etc/stackstate-agent/ /var/log/stackstate-agent/ \
+ && chmod 755 /probe.sh /init-trace.sh
+
+
+#   - copy default config files
```
Comment (TizianoPerrucci) [inline Dockerfiles/trace-agent/Dockerfile]: `# Configuration`

File: Dockerfiles/trace-agent/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+COPY stackstate*.yaml etc/stackstate-agent/
+
+
+# Expose DogStatsD and trace-agent ports
```
Comment (TizianoPerrucci) [inline Dockerfiles/trace-agent/Dockerfile]: `Expose trace-agent port`

File: Dockerfiles/trace-agent/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+####################################
+# Actual docker image construction #
+####################################
+
+FROM debian:buster-slim
+LABEL maintainer "StackState <info@stackstate.com>"
+ENV DOCKER_STS_AGENT=true \
+    PATH=/opt/stackstate-agent/bin/agent/:/opt/stackstate-agent/embedded/bin/:$PATH \
+    CURL_CA_BUNDLE=/opt/stackstate-agent/embedded/ssl/certs/cacert.pem \
+    #Set API Key
```
Comment (TizianoPerrucci) [inline Dockerfiles/trace-agent/Dockerfile]: remove default STS_API_KEY

File: Dockerfiles/trace-agent/entrypoint/init-trace.sh [hunk 12]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
```
Comment (TizianoPerrucci) [inline Dockerfiles/trace-agent/entrypoint/init-trace.sh]: add:

```
if [[ -z "$STS_API_KEY" ]]; then
    echo "You must set an STS_API_KEY environment variable to run the StackState Trace Agent container"
    exit 1
fi
```

File: Dockerfiles/trace-agent/stackstate-docker.yaml [hunk 13]
```diff
@@ -0,0 +1,15 @@
+## Provides autodetected defaults, for vanilla Docker environments,
+## please see stackstate.yaml.example for all supported options
+
+# Autodiscovery settings for vanilla Docker
```
Comment (TizianoPerrucci) [inline Dockerfiles/trace-agent/stackstate-docker.yaml]: try to remove `listeners` and `config_proveder` they should not be needed

File: omnibus/package-scripts/publish_trace_agent.sh [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+cp $ARTIFACT_PATH Dockerfiles/trace-agent
+
+docker build -t stackstate/${IMAGE_REPO}:${IMAGE_TAG} Dockerfiles/trace-agent
+
+
+docker login -u $DOCKER_USER -p $DOCKER_PASS
+docker push stackstate/${IMAGE_REPO}:${IMAGE_TAG}
+
+if ["$PUSH_LATEST= true" ]; then
```
Comment (TizianoPerrucci) [inline omnibus/package-scripts/publish_trace_agent.sh]: `if [ "$PUSH_LATEST" = "true" ]; then`

File: .gitlab-ci.yml (hunk 1 above)
Comment (johnnygladwin1991) [reply .gitlab-ci.yml]: removed

File: .gitlab-ci.yml (hunk 2 above)
Comment (johnnygladwin1991) [reply .gitlab-ci.yml]: changed

File: .gitlab-ci.yml (hunk 3 above)
Comment (johnnygladwin1991) [reply .gitlab-ci.yml]: removed

File: .gitlab-ci.yml (hunk 4 above)
Comment (johnnygladwin1991) [reply .gitlab-ci.yml]: added the test repo

File: .gitlab-ci.yml (hunk 5 above)
Comment (johnnygladwin1991) [reply .gitlab-ci.yml]: removed and added before_script

File: .gitlab-ci.yml (hunk 8 above)
Comment (johnnygladwin1991) [reply .gitlab-ci.yml]: added

File: .gitlab-ci.yml (hunk 7 above)
Comment (johnnygladwin1991) [reply .gitlab-ci.yml]: removed

File: .gitlab-ci.yml (hunk 6 above)
Comment (johnnygladwin1991) [reply .gitlab-ci.yml]: added release repo

File: Dockerfiles/trace-agent/Dockerfile (hunk 11 above)
Comment (johnnygladwin1991) [reply Dockerfiles/trace-agent/Dockerfile]: removed

File: Dockerfiles/trace-agent/Dockerfile (hunk 9 above)
Comment (johnnygladwin1991) [reply Dockerfiles/trace-agent/Dockerfile]: added comment

File: Dockerfiles/trace-agent/Dockerfile (hunk 10 above)
Comment (johnnygladwin1991) [reply Dockerfiles/trace-agent/Dockerfile]: added comment

File: Dockerfiles/trace-agent/entrypoint/init-trace.sh (hunk 12 above)
Comment (johnnygladwin1991) [reply Dockerfiles/trace-agent/entrypoint/init-trace.sh]: added

File: Dockerfiles/trace-agent/stackstate-docker.yaml (hunk 13 above)
Comment (johnnygladwin1991) [reply Dockerfiles/trace-agent/stackstate-docker.yaml]: removed

File: omnibus/package-scripts/publish_trace_agent.sh (hunk 14 above)
Comment (johnnygladwin1991) [reply omnibus/package-scripts/publish_trace_agent.sh]: changed the comparison to String

**Pull request merged**